packages/context/src/interceptor.ts

```diff
@@ -0,0 +1,432 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {
+  ClassDecoratorFactory,
+  DecoratorFactory,
+  MetadataAccessor,
+  MetadataInspector,
+  MetadataMap,
+  MethodDecoratorFactory,
+} from '@loopback/metadata';
+import * as assert from 'assert';
+import * as debugFactory from 'debug';
+import {Binding, BindingTemplate} from './binding';
+import {filterByTag} from './binding-filter';
+import {BindingAddress} from './binding-key';
+import {Context} from './context';
+import {ContextBindings, ContextTags} from './keys';
+import {transformValueOrPromise, ValueOrPromise} from './value-promise';
+const debug = debugFactory('loopback:context:interceptor');
+const getTargetName = DecoratorFactory.getTargetName;
+
+/**
+ * Array of arguments for a method invocation
+ */
+// tslint:disable-next-line:no-any
+export type InvocationArgs = any[];
+
+/**
+ * Return value for a method invocation
+ */
+// tslint:disable-next-line:no-any
+export type InvocationResult = any;
+
+/**
+ * A type for class or its prototype
+ */
+// tslint:disable-next-line:no-any
+type ClassOrPrototype = any;
+
+/**
+ * InvocationContext represents the context to invoke interceptors for a method.
+ * The context can be used to access metadata about the invocation as well as
+ * other dependencies.
+ */
+export class InvocationContext extends Context {
+  /**
+   * Construct a new instance of `InvocationContext`
+   * @param parent Parent context, such as the RequestContext
+   * @param target Target class (for static methods) or prototype/object
+   * (for instance methods)
+   * @param methodName Method name
+   * @param args An array of arguments
+   */
+  constructor(
+    parent: Context,
+    public readonly target: object,
+    public readonly methodName: string,
+    public readonly args: InvocationArgs,
+  ) {
+    super(parent);
+  }
+
+  /**
+   * Discover all binding keys for global interceptors (tagged by
+   * ContextTags.GLOBAL_INTERCEPTOR)
+   */
+  getGlobalInterceptorBindingKeys(): string[] {
+    const bindings: Readonly<Binding<Interceptor>>[] = this.find(
+      filterByTag(ContextTags.GLOBAL_INTERCEPTOR),
+    );
+    this.sortGlobalInterceptorBindings(bindings);
+    return bindings.map(b => b.key);
+  }
+
+  /**
+   * Sort global interceptor bindings by `globalInterceptorGroup` tags
+   * @param bindings An array of global interceptor bindings
+   */
+  private sortGlobalInterceptorBindings(
```

Comment:
Can this leverage new functionality introduced in #2848?

Reply:
Yes. After #2848 is landed.